control-base/src/main/java/fi/nls/oskari/control/feature/InsertFeatureHandler.java

```diff
@@ -47,100 +26,37 @@
 import javax.xml.parsers.ParserConfigurationException;
 import java.io.ByteArrayInputStream;
 import java.io.IOException;
-import java.util.Set;
 
 @OskariActionRoute("InsertFeature")
-public class InsertFeatureHandler extends ActionHandler {
+public class InsertFeatureHandler extends AbstractFeatureHandler {
     private final static Logger LOG = LogFactory.getLogger(InsertFeatureHandler.class);
 
-    private OskariLayerService layerService;
-    private PermissionsService permissionsService;
-    private WFSLayerConfigurationService layerConfigurationService;
-
-    @Override
-    public void init() {
-        super.init();
-        layerService = new OskariLayerServiceIbatisImpl();
-        permissionsService = new PermissionsServiceIbatisImpl();
-        layerConfigurationService = new WFSLayerConfigurationServiceIbatisImpl();
-    }
-
     @Override
     public void handleAction(ActionParameters params)
             throws ActionException {
-        final JSONObject jsonPayload = getJSONPayload(params);
-        // throws denied exception if user doesn't have permission to edit the layer
-        final OskariLayer layer = getLayerForEditing(jsonPayload.optString("layerId"), params.getUser());
-        clearLayerTileCache(layer.getId());
-        final String wfstMessage = createWFSTMessage(jsonPayload, layer);
-        LOG.debug("Inserting feature to service at", layer.getUrl(), "with payload", wfstMessage);
-        final String responseString = postWFSTMessage(layer, wfstMessage);
-        final String updatedFeatureId = parseFeatureIdFromResponse(responseString);
-        ResponseHelper.writeResponse(params, JSONHelper.createJSONObject("fid", updatedFeatureId));
-    }
 
-    private JSONObject getJSONPayload(ActionParameters params)
-            throws ActionException {
+        JSONObject jsonPayload = params.getHttpParamAsJSON("featureData");
+        OskariLayer layer = getLayer(jsonPayload.optString("layerId"));
 
-        String featureData = params.getHttpParam("featureData");
-        JSONObject jsonObject = JSONHelper.createJSONObject(featureData);
-        if (jsonObject == null) {
-            throw new ActionParamsException("Couldn't parse featureData JSON from request");
+        if (!canEdit(layer, params.getUser())) {
```

Comment:
DeleteFeatureHandler checks `params.requireLoggedInUser();` prior to calling this, I like it.